src/bin/server/app.rs

```diff
@@ -0,0 +1,359 @@
+mod test {
+    use super::*;
+    use crate::attestations::get_message_digest;
+    use crate::attestations::{
+        AggregatePriceIntervalEntry, OracleMessage, PriceIntervalEntry, PriceValueEntry,
+    };
+    use axum::{body::Body, http::Request};
+    use bls::{AggregateSignature, SecretKey, Signature};
+    use bytes::Bytes;
+    use tower::ServiceExt;
+
+    async fn get_db_pool() -> SqlitePool {
+        let test_db_url = "sqlite::memory:";
+        let pool = SqlitePool::connect(test_db_url).await.unwrap();
+        sqlx::migrate!("./migrations")
+            .run(&pool)
+            .await
+            .expect("Failed to migrate database");
+        pool
+    }
+
+    enum TestRequest {
+        Get(),
+        Post(Body),
+    }
+    struct TestApp {
+        db_pool: SqlitePool,
+    }
+
+    impl TestApp {
+        pub async fn new() -> Self {
+            let db_pool = get_db_pool().await;
+            TestApp { db_pool }
+        }
+
+        pub async fn get(&self, uri: &str, expected_code: u16) -> Bytes {
+            self.send_request(TestRequest::Get(), uri, expected_code)
+                .await
+        }
+
+        pub async fn post(&self, uri: &str, body: Body, expected_code: u16) -> Bytes {
+            self.send_request(TestRequest::Post(body), uri, expected_code)
+                .await
+        }
+
+        async fn send_request(&self, request: TestRequest, uri: &str, expected_code: u16) -> Bytes {
+            let app = get_app_with_db_pool(self.db_pool.clone());
+            let req = match request {
+                TestRequest::Get() => Request::builder()
+                    .uri(uri)
+                    .method("GET")
+                    .body(Body::empty())
+                    .unwrap(),
+                TestRequest::Post(body) => Request::builder()
+                    .uri(uri)
+                    .method("POST")
+                    .header("Content-Type", "application/json")
+                    .body(body)
+                    .unwrap(),
+            };
+
+            let response = app.oneshot(req).await.unwrap();
+            assert_eq!(response.status().as_u16(), expected_code);
+            let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+            body
+        }
+    }
+
+    fn get_test_message() -> OracleMessage {
+        let test_data_file = std::fs::File::open("./test_data/input/17292025.json").unwrap();
+        serde_json::from_reader(test_data_file).unwrap()
+    }
+
+    fn sign_oracle_message_with_new_key(
+        mut message: OracleMessage,
+        private_key: &SecretKey,
+    ) -> OracleMessage {
+        message.validator_public_key = private_key.public_key();
+        message.value_message.signature =
+            sign_message(&message.value_message.message, &private_key);
+        for interval_message in message.interval_inclusion_messages.iter_mut() {
+            interval_message.signature = sign_message(&interval_message.message, &private_key);
```

Comment:
Don't feel strongly about this but this line makes me feel a [builder pattern](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html) may be healthy. I'm not sure this is the case here, but if `OracleMessage` is complex to construct, where a valid message is only one with a signature, and yet you somehow need to pass messages without a signature, it could be healthy to have a builder hold that half-baked state, and then you allow others to clone that, but spitting out an OracleMessage from a builder is only possible if you provide a key so that it is signed. This way you can have a half built OracleMessage, yet never worry that a passed OracleMessage is not actually "valid" because it somehow ended up lacking a signature.

For practical reasons Builder may be overkill here in which case feel free to ignore but I do like the above pattern when I want to make it as hard as possible for future users of my code to construct invalid values.

Reply:
Mhm, this is only test code so not sure if we need to worry too much about the interface here. 
In general there should not be any need to construct OracleMessages in the server code (except for test purposes).

In general I was thinking about extracting all of the OracleMessage logic into a separate lib / crate that can be used / imported by both client and server.
This would:

1. Ensure that both use compatible logic ( for example for serializing / deserializing / verifying) messages
2. Reduce code duplication (currently the whole struct definition is duplicated in both repos).

What do you think about that ?
If so we could add a builder class in that context if we feel it would be of help.